src/engine.rs

```diff
@@ -0,0 +1,2625 @@
+            wobble_weighted_pos_sum: (0.0, 0.0),
+            wobble_distance_sum: 0.0,
+            position_modeler: None,
+            state_modeler: StateModeler::new(params.stylus_state_modeler_max_input_samples),
+        }
+    }
+
+    /// Clears any in-progress stroke, keeping the same model parameters
+    /// Technically the error is not needed anymore as the params are kept
+    /// as is and can't be uninitialized
+    pub fn reset(&mut self) -> Result<(), i32> {
+        self.last_event = None;
+        self.wobble_deque = VecDeque::with_capacity(
+            (2.0 * self.params.sampling_min_output_rate * self.params.wobble_smoother_timeout)
+                as usize,
+        );
+        self.wobble_duration_sum = 0.0;
+        self.wobble_weighted_pos_sum = (0.0, 0.0);
+        self.position_modeler = None;
+        Ok(()) // to match
+    }
+
+    /// Clears any in-progress stroke, and re initialize the model with
+    /// the given parameters
+    ///
+    /// Here the error is also obsolete as the `ModelerParams` is expected
+    /// to have been built with [ModelerParams::validate] that validates the
+    /// parameters
+    pub fn reset_w_params(&mut self, params: ModelerParams) -> Result<(), i32> {
+        match params.validate() {
+            Ok(params) => self.params = params,
+            Err(_) => return Err(0),
+        };
+        self.last_event = None;
+        self.wobble_deque = VecDeque::with_capacity(
+            (2.0 * self.params.sampling_min_output_rate * self.params.wobble_smoother_timeout)
+                as usize,
+        );
+        self.wobble_duration_sum = 0.0;
+        self.wobble_weighted_pos_sum = (0.0, 0.0);
+        self.wobble_distance_sum = 0.0;
+        self.position_modeler = None;
+        Ok(())
+    }
+
+    /// Updates the model with a raw input, and appends newly generated Results to the results vector.
+    /// Any previously generated Result values remain valid.
+    /// (This does not require that any previous results returned remain in the results vector, as it is
+    /// appended to without examining the existing contents)
+    ///
+    /// If this does not return an error, results will contain at least one Result, and potentially
+    /// more if the inputs are slower than the minimum output rate
+    ///
+    /// for now rnote's wrapper codes verify that the input is not duplicated and time increases between strokes
+    /// This is not tested here, as we suppose that these things are verified beforehand
+    pub fn update(&mut self, input: ModelerInput) -> Result<Vec<ModelerResult>, i32> {
+        match input.event_type {
+            ModelerInputEventType::kDown => {
+                // assumed this is the first ever event
+                self.wobble_update(&input); // first event is "as is"
+
+                // create the position modeler
+                self.position_modeler = Some(PositionModeler::new(self.params, input.clone()));
```

Comment:
Can there be a check that the first event is `Down`, because otherwise users only get a unexplained panic when the modeler is unwrapped and it might not be clear what the issue is.

Reply:
I've added this in the error validation part.
This means that an error occurs if a Down event occurs whilst a stroke is in progress.